subsys/lorawan/lorawan.c

```diff
@@ -0,0 +1,373 @@
+#include <zephyr.h>
+
+#include <LoRaMac.h>
+
+#ifdef CONFIG_LORAMAC_REGION_AS923
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_AS923
+#elif CONFIG_LORAMAC_REGION_AU915
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_AU915
+#elif CONFIG_LORAMAC_REGION_CN470
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_CN470
+#elif CONFIG_LORAMAC_REGION_CN779
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_CN779
+#elif CONFIG_LORAMAC_REGION_EU433
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_EU433
+#elif CONFIG_LORAMAC_REGION_EU868
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_EU868
+#elif CONFIG_LORAMAC_REGION_KR920
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_KR920
+#elif CONFIG_LORAMAC_REGION_IN865
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_IN865
+#elif CONFIG_LORAMAC_REGION_US915
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_US915
+#elif CONFIG_LORAMAC_REGION_RU864
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_RU864
+#elif
+	#error "Atleast one LoRaWAN region should be selected"
+#endif
+
+#define LOG_LEVEL CONFIG_LORAWAN_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(lorawan);
+
+K_SEM_DEFINE(lorawan_config_sem, 0, 1);
+K_SEM_DEFINE(lorawan_tx_sem, 0, 1);
+
+K_MUTEX_DEFINE(lorawan_join_mutex);
+K_MUTEX_DEFINE(lorawan_send_mutex);
+
+static volatile bool join_status = false;
+static volatile bool send_status = false;
+
+/* MAC status strings */
+const char *to_status_str[] = {
+	"OK",				/* LORAMAC_STATUS_OK */
+	"Busy",				/* LORAMAC_STATUS_BUSY */
+	"Service unknown",		/* LORAMAC_STATUS_SERVICE_UNKNOWN */
+	"Parameter invalid",		/* LORAMAC_STATUS_PARAMETER_INVALID */
+	"Frequency invalid",		/* LORAMAC_STATUS_FREQUENCY_INVALID */
+	"Datarate invalid",		/* LORAMAC_STATUS_DATARATE_INVALID */
+	"Frequency or datarate invalid",/* LORAMAC_STATUS_FREQ_AND_DR_INVALID */
+	"No network joined",		/* LORAMAC_STATUS_NO_NETWORK_JOINED */
+	"Length error",			/* LORAMAC_STATUS_LENGTH_ERROR */
+	"Region not supported",		/* LORAMAC_STATUS_REGION_NOT_SUPPORTED */
+	"Skipped APP data",		/* LORAMAC_STATUS_SKIPPED_APP_DATA */
+	"Duty-cycle restricted",	/* LORAMAC_STATUS_DUTYCYCLE_RESTRICTED */
+	"No channel found",		/* LORAMAC_STATUS_NO_CHANNEL_FOUND */
+	"No free channel found",	/* LORAMAC_STATUS_NO_FREE_CHANNEL_FOUND */
+	"Busy beacon reserved time",	/* LORAMAC_STATUS_BUSY_BEACON_RESERVED_TIME */
+	"Busy ping-slot window time",	/* LORAMAC_STATUS_BUSY_PING_SLOT_WINDOW_TIME */
+	"Busy uplink collision",	/* LORAMAC_STATUS_BUSY_UPLINK_COLLISION */
+	"Crypto error",			/* LORAMAC_STATUS_CRYPTO_ERROR */
+	"FCnt handler error",		/* LORAMAC_STATUS_FCNT_HANDLER_ERROR */
+	"MAC command error",		/* LORAMAC_STATUS_MAC_COMMAD_ERROR */
+	"ClassB error",			/* LORAMAC_STATUS_CLASS_B_ERROR */
+	"Confirm queue error",		/* LORAMAC_STATUS_CONFIRM_QUEUE_ERROR */
+	"Multicast group undefined",	/* LORAMAC_STATUS_MC_GROUP_UNDEFINED */
+	"Unknown error",		/* LORAMAC_STATUS_ERROR */
+};
+
+/* MAC event info status strings */
+const char *to_event_info_status_str[] = {
```

Comment:
same here

Reply:
> I'm sorry @Mani-Sadhasivam , I'm not quite able to understand what the problem is.

@SebastianBoe Till now, `zephyr_library_named(loramac-node)` is declared by `drivers/lora/CMakeLists.txt` but for LoRaWAN, we need to use library reference in `subsys/lorawan`. If we use `set(ZEPHYR_CURRENT_LIBRARY loramac-node)` in `subsys/lorawan` it doesn't work because this gets processed before `drivers/lora` and the library definition is missing.

To solve this issue, I moved the library definition to `subsys/lorawan` and it works well. But for the plain LoRa usecase where we don't need `subsys/lorawan`, this is creating issue because the `loramac-node` library definition is missing as `subsys/lorawan` is not enabled.

For me, the only way to fix this issue is by enabling `CONFIG_LORAWAN` for plain LoRa also but I don't favor it.

Reply:
Can't you create more libraries? One for each entity; driver, subsystem, lorawan, lora ?

Then you wouldn't need to 

set(ZEPHYR_CURRENT_LIBRARY loramac-node)

Reply:
@SebastianBoe But the corresponding library will be used by the loramac-node module to include drivers: https://github.com/zephyrproject-rtos/loramac-node/blob/master/zephyr/CMakeLists.txt#L3

If we use different libraries, how will the loramac-node module add the drivers?

Reply:
Couldn't it also create it's own library?

Reply:
Sorry, I'm confused. If all 3 are different libraries, then how `dirvers/lora` and `subsys/lorawan` will use the drivers present in loramac-node? Am I missing something?

Reply:
I am also confused. What is the problem here?

One system can use another by having it invoke functions in the other system.

Which functions are in which libraries is unimportant. Any library can invoke functions from any other library.

Reply:
Here is the error after making different libraries: https://pastebin.ubuntu.com/p/NT9TBDkcNP/

Reply:
Right, to solve this you would want to do

zephyr_include_directories(path/to/LoRaMac.h)

this is assuming that the header files are fairly unique, if not you need to do something that is a bit more involved.

Reply:
Used this as per Slack conversation: https://github.com/zephyrproject-rtos/zephyr/pull/23664/commits/f3068ee2aa278d72e2453b8b44e2bf8d16c41c4b#diff-4d3f1718e6c473abc54edf7fd4c5830d